---
title: "PRIMED example notebook: import from Terra"
output: html_notebook
---

# Import data to workspace: Terra example

In this notebook we will import 1000 Genomes data into data tables, using the Bioconductor AnVIL package to interact with the workspace.

## Install and load R packages

```{r}
# get the latest version of AnVIL from github
#remotes::install_github("Bioconductor/AnVIL", dependencies=FALSE)
library(AnVIL)
library(tidyverse)
```

We are going to import data from another (open-access) AnVIL workspace.

First, we look at the data tables:

```{r}
avtables(namespace="amp-t2d-op", name="2019_ASHG_Reproducible_GWAS-V2")
```

We import the "sample" table as a tibble:

```{r}
samp <- avtable("sample", namespace="amp-t2d-op", name="2019_ASHG_Reproducible_GWAS-V2")
head(samp)
```

All phenotypes were simulated for the purposes of the GWAS tutorial in the origin workspace. In PRIMED (as in dbGaP and other AnVIL consortia) we distinguish between subjects (participants in a study) and samples (e.g. blood sample for genotyping). Multiple samples may be associated with the same subject. All of the data in this table more appropriately belongs in the subject table. In the PRIMED data model, we will actually have multiple tables, with the subject table containing key information like study and consent, and separate phenotype tables for each phenotype domain.

```{r}
subj <- samp %>%
  rename(subject_id=sample_id, 
         superpopuation=ancestry)
head(subj)
```

The sample table links samples to subjects and identifies the data type of each sample.

```{r}
samp <- samp %>%
  select(sample_id) %>%
  mutate(subject_id=sample_id,
         data_type="sequencing")
head(samp)
```

We now add the subject and sample tables to our workspace.

```{r}
avworkspace_namespace()
avworkspace_name()
avtable_import(subj, entity="subject_id")
avtable_import(samp, entity="sample_id")
```

```{r}
avtables()
```

Now we add data files for these samples. In the source workspace, the paths to the VCF files are stored in a "Workspace data" table.

```{r}
avdata(namespace="amp-t2d-op", name="2019_ASHG_Reproducible_GWAS-V2")
```

```{r}
vcf <- avdata(namespace="amp-t2d-op", name="2019_ASHG_Reproducible_GWAS-V2") %>%
  filter(key == "vcf_files") %>%
  select(value) %>%
  unlist()
head(vcf)
```

```{r}
vf <- strsplit(vcf, split='\",\"', fixed=TRUE) %>%
  unlist(use.names=FALSE)

vf <- sub('[\"', '', vf, fixed=TRUE)
vf <- sub('\"]', '', vf, fixed=TRUE)

vf
```
Extract chromosome from the file name

```{r}
chr <- stringr::str_extract(vf, "chr[:alnum:]+[:punct:]")
chr <- sub("chr", "", chr, fixed=TRUE)
chr <- sub(".", "", chr, fixed=TRUE)
vcf <- tibble(chr=chr, file=vf)
```
