---
title: "TOPMed data"
output: html_notebook
---

# Preparing data tables for TOPMed data

## Copy sample and subject files to local instance

```{r}
library(AnVIL)
library(dplyr)
library(readr)
library(jsonlite)
```

```{r}
avtables()
```

```{r}
avtable("reference_file")
```

Use one of the text files to get the list of samples in the freeze9b VCF files.

```{r}
txt_files <- avtable("reference_file") %>%
  filter(`pfb:data_format` == "TXT", grepl("TOPMed_WGS_genetic_sex.txt$", `pfb:file_name`)) %>%
  select(`pfb:ga4gh_drs_uri`) %>%
  unlist(use.names = FALSE)
```

```{r}
#drs_cp(txt_files, ".") can't use for controlled access

system("pip install --upgrade --no-cache-dir terra-notebook-utils")
tnu <- "/home/rstudio/.local/bin/tnu"
system(paste(tnu, "config set-workspace", avworkspace()))
system(paste(tnu, "config set-workspace-namespace", avworkspace_namespace()))
system(paste(tnu, "drs copy-batch", paste(txt_files, collapse=" "), "--dst ."))
```

```{r}
list.files(pattern="*.txt")
```
```{r}
idfile <- "ARIC_phs001211_TOPMed_WGS_genetic_sex.txt"
f9samp <- read_tsv(idfile, show_col_types = FALSE)$SAMPLE_ID
```


CC uploaded the current versions of the subject and sample files from dbGaP.

```{r}
new_file_names <- c(
  "phs001211.v4.pht005754.v4.p3.TOPMed_WGS_ARIC_Subject.MULTI.txt.gz",
  "phs001211.v4.pht005755.v4.p3.TOPMed_WGS_ARIC_Sample.MULTI.txt.gz",
  "phs001211.v4.pht005757.v3.p3.c1.TOPMed_WGS_ARIC_Sample_Attributes.HMB-IRB.txt.gz"
)
new_file_paths <- file.path(avbucket(), "uploaded_data_cc", new_file_names)
gsutil_cp(new_file_paths, ".")
```


```{r}
files <- list(
    subj="phs001211.v4.pht005754.v4.p3.TOPMed_WGS_ARIC_Subject.MULTI.txt.gz",
    samp="phs001211.v4.pht005755.v4.p3.TOPMed_WGS_ARIC_Sample.MULTI.txt.gz",
    attr="phs001211.v4.pht005757.v3.p3.c1.TOPMed_WGS_ARIC_Sample_Attributes.HMB-IRB.txt.gz"
)

dat <- lapply(files, read_tsv, col_types=cols(.default=col_character()), comment="#")
```


## Create tables for PRIMED data model

```{r}
dat <- dat$attr %>%
  left_join(dat$samp) %>%
  left_join(dat$subj) %>%
  filter(SAMPLE_ID %in% f9samp)
```

```{r}
proj <- avtable("project")
phs <- substr(proj$`pfb:dbgap_accession_number`,1,9)
str <- strsplit(proj$`pfb:project_name`, "_") %>% unlist()
study = str[1]
cons <- str[2]
```

```{r}
subject_table <- dat %>%
  select(subject_id = SUBJECT_ID,
         dbgap_subject_id = dbGaP_Subject_ID) %>%
  distinct() %>%
  mutate(consent_code = cons,
         study_nickname = study,
         dbgap_submission = TRUE,
         dbgap_study_id = phs,
         reported_sex = NA) # sometimes this is in subj file
```

```{r}
count(dat, BODY_SITE)
```

```{r}
count(dat, TOPMed_Phase, TOPMed_Project, Funding_Source, SEQUENCING_CENTER)
```

```{r}
sample_table <- dat %>%
  mutate(batch = ifelse(!is.na(TOPMed_Phase),
                        paste0(SEQUENCING_CENTER, "_Phase", TOPMed_Phase),
                        paste0(SEQUENCING_CENTER, "_", Funding_Source))) %>%
  select(sample_id = SAMPLE_ID,
         subject_id = SUBJECT_ID,
         dbgap_sample_id = dbGaP_Sample_ID,
         batch) %>%
  mutate(tissue_source = "UBERON:0000178") # UBERON code for whole blood
```

```{r}
sample_set <- dat %>%
  select(sample_id = SAMPLE_ID) %>%
  mutate(sample_set_id = paste("topmed", "freeze9", study, cons, sep="_"))
```

```{r}
fields <- list(
  sample_set_id = paste("topmed", "freeze9", study, cons, sep="_"),
  seq_center = paste0(unique(dat$SEQUENCING_CENTER), collapse=" | "),
  reference_assembly = "GRCh38",
  alignment_method = "BWA-MEM",
  sequencing_assay = "WGS",
  seq_platform = "Illumina HiSeq X"
)
dataset_table <- tibble(field=names(fields),
                        value=unlist(fields))
```

```{r}
chr_from_filename <- function(f) {
  chr <- stringr::str_extract(f, "chr[:alnum:]+[:punct:]")
  chr <- sub("chr", "", chr, fixed=TRUE)
  chr <- sub(".", "", chr, fixed=TRUE)
  return(chr)
}

file_table_1 <- avtable("reference_file") %>%
  filter(`pfb:data_format` == "VCF") %>%
  mutate(chromosome = chr_from_filename(`pfb:file_name`)) %>%
  select(file_path = `pfb:ga4gh_drs_uri`,
         file_name = `pfb:file_name`,
         file_type = `pfb:data_format`,
         md5sum = `pfb:md5sum`,
         chromosome)
```

```{r}
file_table_2 <- avtable("reference_file") %>%
  filter(`pfb:data_format` == "TXT", grepl("TOPMed_WGS", `pfb:file_name`)) %>%
  select(file_path = `pfb:ga4gh_drs_uri`,
         file_name = `pfb:file_name`,
         md5sum = `pfb:md5sum`) %>%
  mutate(file_type = ifelse(grepl("_DD.txt$", file_name), "data dictionary", 
                     ifelse(grepl("kinship", file_name), "supporting file",
                            "quality metrics"))) %>%
  mutate(chromosome = NA)

file_table <- bind_rows(file_table_1, file_table_2)
```

```{r}
bucket <- avbucket()

tables <- list("subject"=subject_table, "sample"=sample_table, "sample_set"=sample_set, 
               "sequencing_dataset"=dataset_table, "sequencing_file"=file_table)
file_list <- list()
for (t in names(tables)) {
  outfile <- paste0(t, "_table.tsv")
  write_tsv(tables[[t]], outfile)
  bucketfile <- paste0(bucket, "/data_tables/", outfile)
  gsutil_cp(outfile, bucketfile)
  file_list[[t]] <- bucketfile
}
```

```{r}
json <- list("validate_genotype_model.table_files" = file_list,
             "validate_genotype_model.model_url" = "https://raw.githubusercontent.com/UW-GAC/primed_data_models/main/PRIMED_genotype_data_model.json",
             "validate_genotype_model.workspace_name" = avworkspace_name(),
             "validate_genotype_model.workspace_namespace" = avworkspace_namespace(),
             "validate_genotype_model.import_tables" = "true"
) %>% toJSON(pretty=TRUE, auto_unbox=TRUE, unbox=TRUE)
write(json, "validate_genotype_model.json")
```
